metric/metrictest/meter.go

```diff
@@ -115,21 +134,30 @@ func (m *MeterImpl) doRecordSingle(ctx context.Context, labels []attribute.KeyVa
 	}})
 }
 
-func NewMeterProvider() (*MeterImpl, metric.MeterProvider) {
+func NewMeterProvider() *MeterProvider {
+	return &MeterProvider{}
+}
+
+func (p *MeterProvider) Meter(name string, opts ...metric.MeterOption) metric.Meter {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	cfg := metric.NewMeterConfig(opts...)
 	impl := &MeterImpl{
+		library: Library{
+			InstrumentationName:    name,
+			InstrumentationVersion: cfg.InstrumentationVersion(),
+			SchemaURL:              cfg.SchemaURL(),
+		},
+		provider:         p,
 		asyncInstruments: internalmetric.NewAsyncInstrumentState(),
 	}
-	return impl, registry.NewMeterProvider(impl)
-}
-
-func NewMeter() (*MeterImpl, metric.Meter) {
-	impl, p := NewMeterProvider()
-	return impl, p.Meter("mock")
+	p.impls = append(p.impls, impl)
+	return metric.WrapMeterImpl(impl)
 }
 
 func (m *MeterImpl) NewSyncInstrument(descriptor metric.Descriptor) (metric.SyncImpl, error) {
-	m.lock.Lock()
-	defer m.lock.Unlock()
+	m.provider.lock.Lock()
+	defer m.provider.lock.Unlock()
```

Comment:
Hah, that wasn't doing anything, but based on your feedback I cleaned up all the locking in this file. The MeterProvider lock protects the batch of results, and I've added a couple of MeterProvider methods to avoid the `m.provider.lock` pattern.